### PR TITLE
Make fill_nothing take an empty string

### PR DESCRIPTION
**Pull request by AdRiley**
### Pull Request Description

I think fill_nothing should work with an empty string. Currently TextType.preciseTypeForValue(arg.asString()) throws if you give it an empty string so this fix avoid that.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Review by jdunkerley — APPROVED** [review]

File: std-bits/table/src/main/java/org/enso/table/data/column/storage/StringStorage.java [hunk 1]
```diff
@@ -52,7 +52,9 @@ public TextType getType() {
   public Storage<?> fillMissing(
       Value arg, StorageType commonType, ProblemAggregator problemAggregator) {
     if (arg.isString()) {
-      TextType newType = TextType.maxType(type, TextType.preciseTypeForValue(arg.asString()));
+      String strArg = arg.asString();
+      TextType newType =
+          strArg.isEmpty() ? type : TextType.maxType(type, TextType.preciseTypeForValue(strArg));
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/data/column/storage/StringStorage.java]: I don't think this logic is correct.

See the following example:
```
t = Table.new [["X", ["abc", Nothing]]] . cast "X" (Value_Type.Char size=3 variable_length=False)
c1 = t.at "X"
c1.value_type == (Value_Type.Char size=3 variable_length=False)
c2 = c1.fill_nothing ""
c2.value_type == ???
```

The cast is correct - all non-null values are of length 3 and are fixed length.

What is the `value_type` after the operation?

Logically, it should become `(Value_Type.Char size=3 variable_length=True)` - meaning that we have strings of _max length_ 3 but they may have variable length (because our content contains lengths 0 and 3).

But I suspect, reading this logic, that the response would be `(Value_Type.Char size=3 variable_length=False)`, because we just return the `type` unchanged `if strArg.isEmpty()`.

-----

Oh actually, reading more into the code, I expect it will throw `ValueTypeMismatchException` when the `""` value is checked in the `StringBuilder`.

----

Please add a test case where we have a fixed-length string column with NULLs and `fill_nothing ""` it. Let's ensure that it succeeds and the type is as expected (`(Value_Type.Char size=3 variable_length=True)`).

**Review by radeusgd — CHANGES_REQUESTED** [review]
I think an important edge case is unhandled, and I suspect the operation will crash if filling a fixed-length string column with `""`.

Please add a test for this case to ensure it works correctly.

**Review by GregoryTravis — APPROVED** [review]

**Comment by AdRiley** [comment]
I'm wondering if fill_nothing shouldn't be changing types at all.

**Comment by radeusgd** [comment]
> I'm wondering if fill_nothing shouldn't be changing types at all.

I assume the alternative is to ensure the `default` value fits the target type, right?

i.e. for a column `(Value_Type.Char size=3 variable_length=False)`, a value `"a"` or `""` would by disallowed, only strings of exactly length 3 would be allowed? And we could return an error message that suggests that if you want to have a different size, you should cast to `variable_length=True` first?

**Comment by AdRiley** [comment]
> I assume the alternative is to ensure the `default` value fits the target type, right?

Yes and to error if it doesn't

> i.e. for a column `(Value_Type.Char size=3 variable_length=False)`, a value `"a"` or `""` would by disallowed, only strings of exactly length 3 would be allowed? And we could return an error message that suggests that if you want to have a different size, you should cast to `variable_length=True` first?

Yes but fixed length strings can still hold strings smaller than their size. Right? At least that's how they work in the db world. The fixed/variable size is to do with how they are stored not the data in them.

**Comment by radeusgd** [comment]
> > i.e. for a column `(Value_Type.Char size=3 variable_length=False)`, a value `"a"` or `""` would by disallowed, only strings of exactly length 3 would be allowed? And we could return an error message that suggests that if you want to have a different size, you should cast to `variable_length=True` first?
> 
> Yes but fixed length strings can still hold strings smaller than their size. Right? At least that's how they work in the db world. The fixed/variable size is to do with how they are stored not the data in them.

At least currently in Enso, fixed length means that all string values within that column have this fixed length. Currently, regardless of the type, all strings are stored in the same way (I guess that could change with followups of #8512).

In DBs, shorter strings (in fixed length column) are usually padded, right?

**Comment by AdRiley** [comment]
> In DBs, shorter strings (in fixed length column) are usually padded, right?

I think from the users/functionality point of view then no they are not.

https://www.postgresql.org/docs/current/datatype-character.html

`CREATE TABLE test1 (a character(4));
INSERT INTO test1 VALUES ('ok');
SELECT a, char_length(a) FROM test1; -- (1)

  a   | char_length
------+-------------
 ok   |           2


CREATE TABLE test2 (b varchar(5));
INSERT INTO test2 VALUES ('ok');
INSERT INTO test2 VALUES ('good      ');
INSERT INTO test2 VALUES ('too long');
ERROR:  value too long for type character varying(5)
INSERT INTO test2 VALUES ('too long'::varchar(5)); -- explicit truncation
SELECT b, char_length(b) FROM test2;

   b   | char_length
-------+-------------
 ok    |           2
 good  |           5
 too l |           5`

But if we are talking about how they are stored on disk then yes they are likely padded.

Again from https://www.postgresql.org/docs/current/datatype-character.html
"Values of type character are physically padded with spaces to the specified width n, and are stored and displayed that way. However, trailing spaces are treated as semantically insignificant and disregarded when comparing two values of type character."

> At least currently in Enso, fixed length means that all string values within that column have this fixed length. Currently, regardless of the type, all strings are stored in the same way (I guess that could change with followups of #8512).

I think this will cause us problems if we try to round trip fixed width fields from and to a database. I think we should change this behaviour but not as part of this MR. For now I'm going to add the tests you suggested and see what happens for this fix.




<blockquote><div>PostgreSQL Documentation</div><div><strong><a href="https://www.postgresql.org/docs/16/datatype-character.html">8.3. Character Types</a></strong></div><div>8.3.&nbsp;Character Types # Table&nbsp;8.4.&nbsp;Character Types Name Description character varying(n), varchar(n) variable-length with limit character(n), char(n), bpchar(n) fixed-length, blank-padded bpchar variable …</div></blockquote>

File: std-bits/table/src/main/java/org/enso/table/data/column/storage/StringStorage.java [hunk 2]
```diff
@@ -51,12 +50,7 @@ public TextType getType() {
   @Override
   public Storage<?> fillMissing(
       Value arg, StorageType commonType, ProblemAggregator problemAggregator) {
-    if (arg.isString()) {
-      TextType newType = TextType.maxType(type, TextType.preciseTypeForValue(arg.asString()));
```
Comment (AdRiley) [inline std-bits/table/src/main/java/org/enso/table/data/column/storage/StringStorage.java:55]: The new type has already been previously calculated at https://github.com/enso-org/enso/blob/31a0dcef659afb1a3aaacfc532b8d194fd511dea/distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso#L1129

Comment (AdRiley) [reply std-bits/table/src/main/java/org/enso/table/data/column/storage/StringStorage.java:55]: So don't need to do it again here

File: std-bits/table/src/main/java/org/enso/table/data/column/storage/StringStorage.java (hunk 1 above)
Comment (AdRiley) [reply std-bits/table/src/main/java/org/enso/table/data/column/storage/StringStorage.java]: I've added some more tests and further simplified the code

**Comment by AdRiley** [comment]
> I'm wondering if fill_nothing shouldn't be changing types at all.

I've changed my mind. I like the type expanding where needed.

File: std-bits/table/src/main/java/org/enso/table/data/column/storage/StringStorage.java [hunk 3]
```diff
@@ -51,12 +50,7 @@ public TextType getType() {
...
-      return fillMissingHelper(arg, new StringBuilder(size(), newType));
-    } else {
-      return super.fillMissing(arg, commonType, problemAggregator);
-    }
+    return super.fillMissing(arg, commonType, problemAggregator);
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/data/column/storage/StringStorage.java]: This override seems to now just delegate to the super, shall we remove the override altogether?

Comment (AdRiley) [reply std-bits/table/src/main/java/org/enso/table/data/column/storage/StringStorage.java]: Yes good call. And fillMissingHelper is also un-required now too. Have changed it.

File: test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso [hunk 4]
```diff
@@ -326,6 +326,32 @@ spec setup =
             actual = t.fill_nothing ["col0", "col1"] default
             actual.at "col0" . to_vector . should_equal [0, 1000, 4, 5, 1000, 1000]
             actual.at "col1" . to_vector . should_equal [1000, 200, 1000, 400, 500, 1000]
+        
+        Test.specify "should allow to fill_nothing from larger string and expands type to fit" <|
+            t = table_builder [["col0", ["0", Nothing, "4", "5", Nothing, Nothing]]] . cast "col0" (Value_Type.Char size=1 variable_length=False)
+            actual = t.fill_nothing ["col0"] "ABCDE"
+            actual.at "col0" . to_vector . should_equal ["0", "ABCDE", "4", "5", "ABCDE", "ABCDE"]
+            if setup.is_database.not then
+                actual.at "col0" . value_type . should_equal (Value_Type.Char size=5 variable_length=True)
```
Comment (radeusgd) [inline test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso:335]: Do you think we could check this invariant for Postgres?

File: test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso [hunk 5]
```diff
@@ -326,6 +326,32 @@ spec setup =
...
+            t = table_builder [["col0", ["0", Nothing, "4", "5", Nothing, Nothing]], ["col1", [Nothing, "200", Nothing, "400", "500", Nothing]]]
+            actual = t.fill_nothing ["col0", "col1"] ""
+            actual.at "col0" . to_vector . should_equal ["0", "", "4", "5", "", ""]
+            actual.at "col1" . to_vector . should_equal ["", "200", "", "400", "500", ""]
+
+        Test.specify "should allow to fill_nothing a fixed width from an empty string" <|
+            t = table_builder [["col0", [Nothing, "200", Nothing, "400", "500", Nothing]]] . cast "col0" (Value_Type.Char size=3 variable_length=False)
+            actual = t.fill_nothing ["col0"] ""
+            actual.at "col0" . to_vector . should_equal ["", "200", "", "400", "500", ""]  
+            actual.at "col0" . value_type . should_equal (Value_Type.Char variable_length=True)
```
Comment (radeusgd) [inline test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso:347]: IMHO while this type becomes variable length, it could retain a known max-size:
```suggestion
            actual.at "col0" . value_type . should_equal (Value_Type.Char size=3 variable_length=True)
```

what do you think?

**Review by radeusgd — APPROVED** [review]
Two small test suggestions that I'm not 100% convinced are needed.
Overall it looks OK.

File: test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso (hunk 4 above)
Comment (AdRiley) [reply test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso:335]: I wasn't sure how to specify "don't test this on SQLite". Do we have anywhere else we do this?

File: test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso (hunk 5 above)
Comment (AdRiley) [reply test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso:347]: Yes that would be better wouldn't it? I'll look what that code change would be.

File: test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso (hunk 4 above)
Comment (radeusgd) [reply test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso:335]: 1. One way is to check the prefix which contains the name of the DB: https://github.com/enso-org/enso/blob/develop/test/Table_Tests/src/Common_Table_Operations/Filter_Spec.enso#L85 But I think that's not quite perfect.

2. The preferred way is to use 'feature flags' from test selection, and I think `setup.test_selection.fixed_length_text_columns` is exactly what is needed here: 
https://github.com/enso-org/enso/blob/develop/test/Table_Tests/src/Common_Table_Operations/Conversion_Spec.enso#L76

Comment (AdRiley) [reply test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso:335]: Thank you

**Comment by AdRiley** [comment]
I'm going to merge this one and resolve the 2 suggestions in a second PR as they have some bigger implications that we should discuss.
